cpp/benchmarks/join/generate_input_tables.cuh

```diff
@@ -0,0 +1,275 @@
+ *                                  "build" the hash table in a hash based join implementation.
+ * @param[in] build_tbl_size        number of keys in the build table
+ * @param[out] probe_tbl            The probe table to generate. Usually the larger table used to
+ *                                  probe into the hash table created from the build table.
+ * @param[in] build_tbl_size        number of keys in the build table
+ * @param[in] selectivity           propability with which an element of the probe table is
+ *                                  present in the build table.
+ * @param[in] rand_max              maximum random number to generate. I.e. random numbers are
+ *                                  integers from [0,rand_max].
+ * @param[in] uniq_build_tbl_keys   if each key in the build table should appear exactly once.
+ */
+template<typename key_type, typename size_type>
+void generate_input_tables(
+                           key_type* const build_tbl,
+                           const size_type build_tbl_size,
+                           key_type* const probe_tbl,
+                           const size_type probe_tbl_size,
+                           const double selectivity,
+                           const key_type rand_max,
+                           const bool uniq_build_tbl_keys)
+{
+    // With large values of rand_max the a lot of temporary storage is needed for the lottery. At the expense of not
+    // being that accurate with applying the selectivity an especially more memory efficient implementations would be
+    // to partition the random numbers into two intervals and then let one table choose random numbers from only one
+    // interval and the other only select with selectivity propability from the same interval and from the other in the
+    // other cases.
+
+    static_assert(std::is_signed<key_type>::value, "key_type needs to be signed for lottery to work");
+
+    const int block_size = 128;
+
+    // Maximize exposed parallelism while minimizing storage for curand state
+    int num_blocks_init_build_tbl {-1};
+    CUDA_TRY(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+        &num_blocks_init_build_tbl, init_build_tbl<key_type, size_type>, block_size, 0
+    ));
+
+    int num_blocks_init_probe_tbl {-1};
+    CUDA_TRY(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+        &num_blocks_init_probe_tbl, init_probe_tbl<key_type,size_type>, block_size, 0
+    ));
+
+    int dev_id {-1};
+    CUDA_TRY(cudaGetDevice(&dev_id));
+
+    int num_sms {-1};
+    CUDA_TRY(cudaDeviceGetAttribute(&num_sms, cudaDevAttrMultiProcessorCount, dev_id));
+
+    const int num_states = num_sms * std::max(num_blocks_init_build_tbl, num_blocks_init_probe_tbl) * block_size;
+    curandState *devStates;
+    CUDA_TRY(cudaMalloc(&devStates, num_states * sizeof(curandState)));
+
+    init_curand<<<(num_states - 1) / block_size + 1, block_size>>>(devStates, num_states);
+
+    CUDA_TRY(cudaGetLastError());
```

Comment:
Use CHECK_STREAM from cuDF instead of these two lines (only synchronizes in debug mode)

Reply:
Resolved in https://github.com/rapidsai/cudf/pull/2105/commits/10d4e20c9ab010c429d853125fb4cffe821554fd.